File: src/jp_quality_gate/gate.py
from __future__ import annotations

from pathlib import Path

from .cj import CJDetector
from .models import GateResult, Issue
from .text import mask_markdown_nonprose
from .unihan import UnihanScanner


class QualityGate:
    def __init__(self, *, unihan_table: Path, cj_detector: CJDetector | None = None) -> None:
        self.unihan = UnihanScanner(unihan_table)
        self.cj = cj_detector or CJDetector()

    def check(
        self,
        text: str,
        *,
        include_code: bool = False,
        cj_min_cjk: int = 4,
        cj_min_gap: float = 0.15,
        warnings_as_errors: bool = False,
    ) -> GateResult:
        masked = mask_markdown_nonprose(text, include_code=include_code)
        issues: list[Issue] = []
        issues.extend(self.unihan.scan(text, masked))
        issues.extend(
            self.cj.scan(
                text,
                masked,
                min_cjk=cj_min_cjk,
                min_gap=cj_min_gap,
            )
        )

        if warnings_as_errors:
            for issue in issues:
                if issue.severity == "warning":
                    issue.severity = "error"

        issues.sort(key=lambda issue: (issue.start, 0 if issue.severity == "error" else 1))
        return GateResult(
            issues=issues,
            meta={
                "unicode_version": self.unihan.unicode_version,
                "cjclassifier_version": getattr(self.cj, "package_version", "injected"),
                "cj_min_cjk": cj_min_cjk,
                "cj_min_gap": cj_min_gap,
                "include_code": include_code,
            },
        )

File: internal/gate/gate.go
// Package gate composes Markdown masking, Unihan scanning, and CJ detection.
package gate

import (
	"sort"

	"github.com/ktutumi/jp-quality-gate/internal/cj"
	"github.com/ktutumi/jp-quality-gate/internal/report"
	jptext "github.com/ktutumi/jp-quality-gate/internal/text"
	"github.com/ktutumi/jp-quality-gate/internal/unihan"
)

// Options controls one quality-gate check.
type Options struct {
	IncludeCode      bool
	CJMinCJK         int
	CJMinGap         float64
	WarningsAsErrors bool
}

// Gate holds the loaded static data reused across a check.
type Gate struct {
	Unihan *unihan.Scanner
	CJ     *cj.Classifier
}

// Check returns findings in Python-compatible source order.
func (g *Gate) Check(input string, options Options) report.GateResult {
	masked := jptext.MaskMarkdownNonProse(input, options.IncludeCode)
	issues := g.Unihan.Scan(input, masked)
	issues = append(issues, g.CJ.Scan(input, masked, options.CJMinCJK, options.CJMinGap)...)

	if options.WarningsAsErrors {
		for index := range issues {
			if issues[index].Severity == report.SeverityWarning {
				issues[index].Severity = report.SeverityError
			}
		}
	}
	sort.SliceStable(issues, func(i, j int) bool {
		if issues[i].Start != issues[j].Start {
			return issues[i].Start < issues[j].Start
		}
		return issues[i].Severity == report.SeverityError && issues[j].Severity != report.SeverityError
	})

	return report.GateResult{
		Issues: issues,
		Meta: map[string]any{
			"unicode_version":      g.Unihan.UnicodeVersion,
			"cjclassifier_version": cj.Version,
			"cj_min_cjk":           options.CJMinCJK,
			"cj_min_gap":           options.CJMinGap,
			"include_code":         options.IncludeCode,
		},
	}
}
